cmd/bucket-targets.go

```diff
@@ -35,14 +36,124 @@ import (
 )
 
 const (
-	defaultHealthCheckDuration = 30 * time.Second
+	defaultHealthCheckDuration = 5 * time.Second
+	// default interval for reload of all remote target endpoints
+	defaultHealthCheckReloadDuration = 30 * time.Minute
 )
 
 // BucketTargetSys represents bucket targets subsystem
 type BucketTargetSys struct {
 	sync.RWMutex
 	arnRemotesMap map[string]*TargetClient
 	targetsMap    map[string][]madmin.BucketTarget
+	hMutex        sync.RWMutex
+	hc            map[string]*healthChecker
+}
+
+// healthChecker struct represents health check of a replication target endpoint.
+type healthChecker struct {
+	Endpoint string
+	Client   *madmin.AnonymousClient
+	Online   bool
+	ctx      context.Context
+	CancelFn context.CancelFunc
+}
+
+// isOffline returns current liveness result of remote target. Lazily start
+// healthcheck if remote endpoint is not present in `hc` map
+func (sys *BucketTargetSys) isOffline(ep string, secure bool) bool {
+	sys.hMutex.RLock()
+	defer sys.hMutex.RUnlock()
+
+	if h, ok := sys.hc[ep]; ok {
+		return !h.Online
+	}
+	go sys.initHC(ep, secure)
+	return false
+}
+
+// initHC initializes an anonymous client for performing health check on the remote endpoint
+// and loads it into `hc` map.
+func (sys *BucketTargetSys) initHC(ep string, secure bool) {
+	clnt, e := madmin.NewAnonymousClient(ep, secure)
+	if e != nil {
+		logger.LogOnceIf(GlobalContext, fmt.Errorf("WARNING: Unable to initialize health check client for %s", ep), string(replicationSubsystem))
+		return
+	}
+	clnt.SetCustomTransport(globalRemoteTargetTransport)
+	sys.hMutex.Lock()
+	defer sys.hMutex.Unlock()
+	ctx, cancelFn := context.WithCancel(GlobalContext)
+	sys.hc[ep] = &healthChecker{
+		Client:   clnt,
+		Endpoint: ep,
+		ctx:      ctx,
+		CancelFn: cancelFn,
+		Online:   true,
+	}
+	go sys.heartBeatFn(ctx, ep)
+}
+
+// heartBeatFn performs liveness check on endpoint and maintains online status for endpoint.
+func (sys *BucketTargetSys) heartBeatFn(ctx context.Context, ep string) {
+	hcTimer := time.NewTimer(defaultHealthCheckDuration)
+	defer hcTimer.Stop()
+	for {
+		select {
+		case <-hcTimer.C:
+			var online bool
+			sys.hMutex.Lock()
+			cctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+			defer cancel()
+			for result := range sys.hc[ep].Client.Alive(cctx, madmin.AliveOpts{}, madmin.ServerProperties{
+				Endpoint: ep,
+			}) {
+				online = result.Online
```

Comment:
Also, I think this can be even simpler you can have a single AnonymousClient globally and re-use that for everything there is no reason to reload it and re-attempt to register it. 

Perhaps an optional API must be present in this, let me add it.

Reply:
Okay added `v1.4.18` please use a global AnonymousClient called `madmin.NewAnonymousClientNoEndpoint()`